rfd/0125-dynamic-auto-discovery-config.md

```diff
@@ -0,0 +1,406 @@
+
+## What
+Dynamically configure matchers for the Auto-Discovery service.
+
+## Why
+Teleport Discover will implement a flow where users are able to install a `discovery_service` and configure it without ever leaving the WebApp.
+
+To do so, the `discovery_service` will be deployed with minimal configuration and let the user configure the matchers afterwards.
+
+Although, this RFD is born from this necessity, there's nothing bounding this feature to using the Discover Wizard, and rather make this dynamic configuration available for any `discovery_service`.
+
+Having an almost stateless configuration eases the configuration and deployment of `discovery_service`s, which is essential for adoption.
+Users no longer need to log in to a machine that is running this service to change the matchers, and can do any sort of configuration from the clients (ie WebApp, `tctl`, Terraform Provider and Kubernetes Operator).
+
+## Details
+There will be a new resource - `DiscoveryConfig` - which will be used by the `discovery_service` to create a set of matchers for that particular service.
+
+We'll create relations between those two entities using labels and label matchers.
+This is common for other type of services, eg `db_service` has the `resources.labels` labels matcher which allows users to define which Databases to proxy based on their labels.
+
+### New `DiscoveryConfig` resource
+A new resource will be created to store the desired matchers.
+
+The `spec` schema of the `DiscoveryConfig` contains all the required fields for matching on resources on supported clouds.
+
+Currently, the schema is compose using the following matchers:
+- `lib/config.AWSMatcher`
+- `lib/config.AzureMatcher`
+- `lib/config.GCPMatcher`
+
+In order to use those types, we must first define them as protobuf types and then import them in `lib/config`.
+
+This way, we have a single place where we define the matchers properties.
+They are then used for file reading or for defining `DiscoveryConfig` resources.
+
+Example:
+```yaml
+kind: DiscoveryConfig
+version: v1
+metadata:
+  name: production-resources
+  labels:
+    cloud: aws
+    env: prod
+spec:
+    aws:
+     - types: ["ec2"]
+       regions: ["us-east-1","us-west-1"]
+       tags:
+         "*": "*"
+       install:
+         join_params:
+           token_name:  "aws-discovery-iam-token"
+         script_name: "default-installer"
+       ssm:
+         document_name: "TeleportDiscoveryInstaller"
+    azure:
+    - types: ["aks"]
+      regions: ["eastus", "westus"]
+      subscriptions: ["11111111-2222-3333-4444-555555555555"]
+      resource_groups: ["group1", "group2"]
+      tags:
+        "*": "*"
+    gcp:
+    - types: ["gke"]
+      locations: ["*"]
+      tags:
+        "*": "*"
+      project_ids: ["myproject"]
+```
+
+### Discovery Group
+The Discovery Group property will not be ported to `DiscoveryConfig` because it exists to ensure that multiple `discovery_services` don't step on each other.
```

Comment:
How do we plan to solve the cases where  one DiscoveryService  removes  resource discovered by other DiscoveryService  without `Discovery Group` ? 

For instance: 
Discovery for us-east-1 Handled by DiscoveryServiceA: 
```yaml
kind: DiscoveryConfig
version: v1
metadata:
  name: production-resources-us-east-1
  labels:
    cloud: aws
    region: us-east-1
    env: prod
spec:
    aws:
     - types: ["ec2"]
       regions: ["us-east-1","us-west-1"]
       tags:
         "*": "*"
 ```
Discovery for `us-west-1` Handled by DiscoveryServiceB: 
```yaml
kind: DiscoveryConfig
version: v1
metadata:
  name: production-resources-us-west-1
  labels:
    cloud: aws
    region: us-west-1
    env: prod
spec:
    aws:
     - types: ["ec2"]
       regions: ["us-west-1"]
       tags:
         "*": "*"
```

In this situation DiscoveryService A won't be able to distinguish if own resource and during discovery agent flow will delete Cloud Resource discovered by Discovery Service B.

Reply:
I have the same opinion as @smallinsky .

We end up requiring it because you are allowed to have multiple discovery agents (can discover other objects from other cloud accounts) running and it will break everything.

Reply:
I would argue that it's even simpler

Focusing on Discover:
User clicks on deploy agent
Teleport deploys an agent with a discovery group of X and no matcher
WebUI creates a DiscoveryConfig with X as discovery group and a set of matchers considering their current selected resource
X can be anything as long as it is uniquely used in those two places (possibly an UUID).

---

As for other usages of the `DiscoveryConfig`, we could document it.
I think it kind of makes sense to use the `discovery_group` to anchor all the static and dynamic (`DiscoveryConfig`) matchers.
But I do agree that it's a new concept and, well, we lose consistency between Discovery and DB/App Services.

Reply:
After discussing this approach with Roman, we concluded that it makes sense to use the `discovery_group` to link `DiscoveryConfig` to `discovery_service`.